--- nfs-utils-1.0.9/utils/mount/nfs.man.orig	2007-07-05 14:40:20.000000000 -0400
+++ nfs-utils-1.0.9/utils/mount/nfs.man	2007-07-05 15:01:00.000000000 -0400
@@ -24,15 +24,15 @@ To mount via NFS version 2, use the
 .BR nfs
 file system type and specify
 .BR nfsvers=2 .
-Version 2 is the default protocol version for the
-.BR nfs
-file system type when
-.BR nfsvers=
-is not specified on the mount command.
 To mount via NFS version 3, use the
 .BR nfs
 file system type and specify
 .BR nfsvers=3 .
+Version 3 is the default protocol version for the
+.BR nfs
+file system type when
+.BR nfsvers=
+is not specified on the mount command.
 To mount via NFS version 4, use the
 .BR nfs4
 file system type.
@@ -153,11 +153,19 @@ If the remote host's NFS daemon is not r
 its portmapper, the standard NFS port number 2049 is
 used instead.
 .TP 1.5i
+.TP 1.5i
+.I proto=n
+Mount the NFS filesystem using a specific network protocol
+instead of the default TCP protocol.
+Valid protocol types are
+.IR udp
+and
+.IR tcp .
+.TP 1.5i
 .I mountport=n
 The numeric value of the
 .B mountd
 port.
-.TP 1.5i
 .I mounthost=name
 The name of the host running
 .B mountd .
@@ -266,12 +274,12 @@ Note that there is a performance penalty
 or privacy.
 .TP 1.5i
 .I tcp
-Mount the NFS filesystem using the TCP protocol instead of the
-default UDP protocol.  Many NFS servers only support UDP.
+Mount the NFS filesystem using the TCP protocol.
+This is the default protocol.
 .TP 1.5i
 .I udp
-Mount the NFS filesystem using the UDP protocol.  This
-is the default.
+Mount the NFS filesystem using the UDP protocol
+instead of the default TCP protocol. 
 .TP 1.5i
 .I nordirplus
 Disables NFSv3 READDIRPLUS RPCs. Use this options when
@@ -397,12 +405,12 @@ used instead.
 .TP 1.5i
 .I proto=n
 Mount the NFS filesystem using a specific network protocol
-instead of the default UDP protocol.
-Many NFS version 4 servers only support TCP.
+instead of the default TCP protocol.
 Valid protocol types are
 .IR udp
 and
 .IR tcp .
+Many NFS version 4 servers only support the TCP protocol.
 .TP 1.5i
 .I clientaddr=n
 On a multi-homed client, this
--- nfs-utils-1.0.9/utils/exportfs/exports.man.orig	2007-07-05 14:40:20.000000000 -0400
+++ nfs-utils-1.0.9/utils/exportfs/exports.man	2007-07-05 15:05:57.000000000 -0400
@@ -105,29 +105,16 @@ This option allows the NFS server to vio
 reply to requests before any changes made by that request have been
 committed to stable storage (e.g. disc drive).
 
-Using this option usually improves performance, but at the cost that
-an unclean server restart (i.e. a crash) can cause data to be lost or
+Using this option might improve performance with version 2 only, 
+but at the cost that an unclean server restart (i.e. a crash) 
+can cause data to be lost or
 corrupted.
-
 .TP
 .IR sync
 Reply to requests only after the changes have been committed to stable
 storage (see
 .IR async
 above).
-
-In releases of nfs-utils upto and including 1.0.0, this option was the
-default.  In this and future releases,
-.I sync
-is the default, and
-.I async
-must be explicit requested if needed.
-To help make system adminstrators aware of this change, 'exportfs'
-will issue a warning if neither
-.I sync
-nor
-.I async
-is specified.
 .TP
 .IR no_wdelay
 This option has no effect if
